### DISCUSSION TITLE
The data in useQuery gets update when using queryClient.setQueriesData of useQueryClient(). Shouldn't onSuccess be also updated?

### DISCUSSION THREAD

**Discussion by Parthvsquare** [TanStack/query#5192] [Q&A] [answered]
` const { isLoading, isError, data, error } = useQuery({
    queryKey: ['todos'],
    queryFn: fetchTodoList,
  })
`
the useQuery has one more option onSuccess, should this be also updated? It would be easier to not use unnecessary useEffect just to watch data. 
 ```
 const { isLoading, isError } = useQuery({
    queryKey: ['todos'],
    queryFn: fetchTodoList,
  },onSuccess(data)=> {
  setData(data)
  })
```
Thus eliminating the need of useEffect.  But if we use this approach and need to update the cache after mutation, the onSuccess's data won't be updated. This cause the need of useEffect.

**Comment by TkDodo** [comment, marked as answer]
This is on purpose - onSuccess is tied to a successful run of the QueryFunction. What is `setData` doing? You likely don't need to do that - neither in onSuccess nor in an effect

**Reply by Parthvsquare** [reply]
The data is {} type. So i am asserting it in onSuccess. (data as unknown as AssertType}. I am currently using memo and passing it to table dataSource
